### PR TITLE
TST: centralize and standardize pandas imports

### PR DESCRIPTION
**Pull request by tacaswell**
Added a decorator to handle:

 - importorskip
 - using the correct unit-handler registration function to deal with:
    - pandas not auto-registering
    - pandas moving an renaming the registration function

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant

### CODE REVIEW

**Comment by QuLogic** [comment]
A more pytest-y way would be with a fixture instead of a decorator, I think.

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -5091,11 +5091,26 @@ def test_broken_barh_empty():
     ax.broken_barh([], (.1, .5))
 
 
-def test_pandas_pcolormesh():
-    pd = pytest.importorskip('pandas')
-    from pandas.tseries import converter
-    converter.register()
+def _pandas_wrapper(func):
+    import functools
```
Comment (dstansby) [inline lib/matplotlib/tests/test_axes.py]: Could this get a one line comment along the lines of "Decorator for tests that use pandas"?

**Comment by tacaswell** [comment]
Both comments addressed.

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -5091,11 +5091,22 @@ def test_broken_barh_empty():
     ax.broken_barh([], (.1, .5))
 
 
-def test_pandas_pcolormesh():
+@pytest.fixture
+def pd(request):
+    '''fixuture to import and configure pandas'''
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: fixture

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -5091,11 +5091,22 @@ def test_broken_barh_empty():
...
     pd = pytest.importorskip('pandas')
-    from pandas.tseries import converter
-    converter.register()
+    if pd:
+        try:
+            from pandas.plotting import (
+                register_matplotlib_converters as register)
+        except ImportError:
+            from pandas.tseries.converter import register
+        register()
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Should you unregister after the test is run? Is that even possible? Or is it covered by the general reset fixture?

File: lib/matplotlib/tests/test_axes.py [hunk 4]
```diff
@@ -5119,10 +5126,8 @@ def test_pandas_indexing_dates():
     ax.plot('dates', 'values', data=without_zero_index)
 
 
-def test_pandas_errorbar_indexing():
-    pd = pytest.importorskip('pandas')
-    from pandas.tseries import converter
-    converter.register()
+def test_pandas_errorbar_indexing(pd):
+    import pandas as pd
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Extra import.

File: lib/matplotlib/tests/test_axes.py [hunk 5]
```diff
@@ -5131,30 +5136,26 @@ def test_pandas_errorbar_indexing():
     ax.errorbar('x', 'y', xerr='xe', yerr='ye', data=df)
 
 
-def test_pandas_indexing_hist():
-    pd = pytest.importorskip('pandas')
-    from pandas.tseries import converter
-    converter.register()
+def test_pandas_indexing_hist(pd):
+    import pandas as pd
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Extra import.

File: lib/matplotlib/tests/test_axes.py [hunk 6]
```diff
@@ -5131,30 +5136,26 @@ def test_pandas_errorbar_indexing():
...
     axes.hist(ser_2)
 
 
-def test_pandas_bar_align_center():
+def test_pandas_bar_align_center(pd):
     # Tests fix for issue 8767
-    pd = pytest.importorskip('pandas')
-    from pandas.tseries import converter
-    converter.register()
+    import pandas as pd
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Extra import.

File: lib/matplotlib/tests/test_axes.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_axes.py]: For newer versions of pandas we can.  Update coming.

**Comment by tacaswell** [comment]
Third times the charm?

File: lib/matplotlib/testing/conftest.py [hunk 7]
```diff
@@ -76,3 +76,26 @@ def mpl_image_comparison_parameters(request, extension):
...
+            from pandas.plotting import (
+                register_matplotlib_converters as register)
+        except ImportError:
+            from pandas.tseries.converter import register
+        register()
+
+        try:
+            from pandas.plotting import (
+                deregister_matplotlib_converters as deregister)
+            request.addfinalizer(deregister)
```
Comment (anntzer) [inline lib/matplotlib/testing/conftest.py]: a yield fixture (in the style of a context manager) seems mildly more pythonic (https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code).

Comment (anntzer) [reply lib/matplotlib/testing/conftest.py]: @tacaswell let us know if you want to make the change or I can make it in a separate pr too

Comment (tacaswell) [reply lib/matplotlib/testing/conftest.py]: @anntzer sorry, been traveling, feel free to push to this branch.

**Pull request merged**